pkg/agentctl/samples.go

```diff
@@ -0,0 +1,131 @@
+package agentctl
+
+import (
+	"time"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/tsdb/record"
+	"github.com/prometheus/prometheus/tsdb/wal"
+)
+
+type Series struct {
+	Labels labels.Labels
+	MinTS  time.Time
+	MaxTS  time.Time
+	Count  int64
+}
+
+// FindSamples TODO
+func FindSamples(walDir string, selectorStr string) ([]*Series, error) {
+	w, err := wal.Open(nil, walDir)
+	if err != nil {
+		return nil, err
+	}
+
+	selector, err := parser.ParseMetricSelector(selectorStr)
+	if err != nil {
+		return nil, err
+	}
+
+	labelsByRef := make(map[uint64]labels.Labels)
+
+	minTSByRef := make(map[uint64]int64)
+	maxTSByRef := make(map[uint64]int64)
+	sampleCountByRef := make(map[uint64]int64)
+
+	// get the references matching label selector
+	err = walIterate(w, func(r *wal.Reader) error {
+		return collectSeries(r, selector, labelsByRef)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// find related samples
+	err = walIterate(w, func(r *wal.Reader) error {
+		return collectSamples(r, labelsByRef, minTSByRef, maxTSByRef, sampleCountByRef)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var series []*Series
+	for ref, labels := range labelsByRef {
+		series = append(
+			series,
+			&Series{
+				Labels: labels,
+				Count:  sampleCountByRef[ref],
+				MinTS:  timestamp.Time(minTSByRef[ref]),
+				MaxTS:  timestamp.Time(maxTSByRef[ref]),
+			},
+		)
```

Comment:
Personally I find the indentation a little weird here, can we change it?

```suggestion
		series = append(series, &Series{
			Labels: labels,
			Count:  sampleCountByRef[ref],
			MinTS:  timestamp.Time(minTSByRef[ref]),
			MaxTS:  timestamp.Time(maxTSByRef[ref]),
		})
```

Reply:
Yes that's better :slightly_smiling_face: